Sources/TuistKit/Commands/TuistCommand.swift

```diff
@@ -74,6 +73,24 @@ public struct TuistCommand: ParsableCommand {
         }
     }
 
+    private static func execute(_ command: ParsableCommand) throws {
+        var command = command
+        let userOptedOutFromStats = Environment.shared.tuistVariables["TUIST_STATS_OPT_OUT"] != nil
+        guard userOptedOutFromStats else {
+            try command.run()
+            return
+        }
+        let commandCompletionGroup = DispatchGroup()
```

Comment:
I'd make concurrency-related code an implementation detail of the async-queue foundation. Otherwise, you end up with logic like the one below that has to make assumptions about what Tuist's async-queue does internally. Moreover, as a developer adding analytics code, I would not expect having to use dispatch groups to control the execution of Tuist.

Reply:
The `AsyncQueue` api has a completion block that informs a consumer when the async task got persisted. 
If we want this to be a private impl detail of the queue, we have to block to thread of the consumer of the `AsyncQueue`, which I think is a bad practice and would still result in "waiting".

The Tuist process runs and exits after the command is executed, if we want it to make sure that the event got persisted before exiting, we have two solutions:
1) waiting for the completion block to be called, and having a parachute timeout (which I set to 2 seconds even though persisting an event won't take more than a hundred ms) in case things go wrong. You can implement that in many ways that are conceptually the same. I choose `DispatchGroup`
2) changing TuistCommand to be reactive, and exit once all relevant events completed (command finished, analytics events persited, etc...) which is basically the same as 1 but the code would read differently

If you have another solution in your mind that doesn't block a thread and make sure Tuist waits for the event to be persisted, let me know

Reply:
@pepibumur I've renamed `completion` to `didPersistEvent` to make it more clear

Reply:
If you prefer to hide all of this, we can do the `DispatchGroup` thingy inside `AsyncQueue`, but I think it's a bad practice, I would imagine `AsyncQueue` to be asynchronous and not block your thread

Reply:
I don't think there is an assumption of what the queue does internally, the completion block is part of the public API of the `AsyncQueue`

Reply:
I see what you mean now. It was a bit confusing when I read "completion". However, I still think there's some room for improvement in that API. For instance, if we want to report an event from another place in the Tuist's codebase, does that mean that we'd need to write the same dispatch-group code? If so, this is a good candidate to be extracted from this implementation. It does not feel it fits here.

Reply:
Ok, I see what you mean, I can extract this into a class that gets invoked before Tuist process is about to exit, and check if we have to wait for some events.